routes/stake.go

```diff
@@ -0,0 +1,417 @@
+	decoder := json.NewDecoder(io.LimitReader(req.Body, MaxRequestBodySizeBytes))
+	requestData := UnlockStakeRequest{}
+	if err := decoder.Decode(&requestData); err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("UnlockStake: Problem parsing request body: %v", err))
+		return
+	}
+
+	// Convert TransactorPublicKeyBase58Check to TransactorPublicKeyBytes
+	if requestData.TransactorPublicKeyBase58Check == "" {
+		_AddBadRequestError(ww, fmt.Sprint("UnlockStake: TransactorPublicKeyBase58Check is required"))
+		return
+	}
+	transactorPublicKeyBytes, _, err := lib.Base58CheckDecode(requestData.TransactorPublicKeyBase58Check)
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("UnlockStake: Problem decoding TransactorPublicKeyBase58Check %s: %v",
+			requestData.TransactorPublicKeyBase58Check, err))
+		return
+	}
+
+	// Convert ValidatorPublicKeyBase58Check to ValidatorPublicKeyBytes
+	if requestData.ValidatorPublicKeyBase58Check == "" {
+		_AddBadRequestError(ww, fmt.Sprint("UnlockStake: ValidatorPublicKeyBase58Check is required"))
+		return
+	}
+	validatorPublicKeyBytes, _, err := lib.Base58CheckDecode(requestData.ValidatorPublicKeyBase58Check)
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("UnlockStake: Problem decoding ValidatorPublicKeyBase58Check %s: %v",
+			requestData.ValidatorPublicKeyBase58Check, err))
+		return
+	}
+
+	// Validate start and end epoch
+	if requestData.StartEpochNumber > requestData.EndEpochNumber {
+		_AddBadRequestError(ww, fmt.Sprint("UnlockStake: StartEpochNumber cannot be greater than EndEpochNumber"))
+		return
+	}
+
+	// Parse ExtraData.
+	extraData, err := EncodeExtraDataMap(requestData.ExtraData)
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("UnlockStake: Problem parsing ExtraData: %v", err))
+		return
+	}
+
+	// Compute the additional transaction fees as specified
+	// by the request body and the node-level fees.
+	additionalOutputs, err := fes.getTransactionFee(
+		lib.TxnTypeUnlockStake,
+		transactorPublicKeyBytes,
+		requestData.TransactionFees,
+	)
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprint("UnlockStake: specified TransactionFees are invalid"))
+		return
+	}
+
+	// Create the transaction.
+	txn, totalInput, changeAmount, fees, err := fes.blockchain.CreateUnlockStakeTxn(
+		transactorPublicKeyBytes,
+		&lib.UnlockStakeMetadata{
+			ValidatorPublicKey: lib.NewPublicKey(validatorPublicKeyBytes),
+			StartEpochNumber:   requestData.StartEpochNumber,
+			EndEpochNumber:     requestData.EndEpochNumber,
+		},
+		extraData,
+		requestData.MinFeeRateNanosPerKB,
+		fes.backendServer.GetMempool(),
+		additionalOutputs,
+	)
+
+	// Construct response.
+	txnBytes, err := txn.ToBytes(true)
+	if err != nil {
+		_AddInternalServerError(ww, fmt.Sprintf("UnlockStake: Problem serializing transaction: %v", err))
+		return
+	}
+
+	res := StakeTxnResponse{
+		SpendAmountNanos:  totalInput - changeAmount - fees,
+		TotalInputNanos:   totalInput,
+		ChangeAmountNanos: changeAmount,
```

Comment:
What do these fields entail in the context of a stake unlock?

Reply:
TotalInputNanos = what's the total amount I'm spending in this transaction
SpendAmountNanos = how much am I spending OTHER than fees
ChangeAmountNanos = 0